src/relay/ir/text_printer.cc

```diff
@@ -0,0 +1,752 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file text_printer.cc
+ * \brief Text printer to print relay in text form.
+ */
+#include <tvm/relay/environment.h>
+#include <tvm/relay/expr_functor.h>
+#include <sstream>
+#include "../pass/type_functor.h"
+#include "../../lang/attr_functor.h"
+
+namespace tvm {
+namespace relay {
+
+/*!
+ * \brief the text value used in text printer.
+ * Defined as a struct for future compatibility reason
+ */
+struct TextValue {
+  /*! \brief The str representation */
+  std::string name;
+  // constructor
+  TextValue() {}
+  // constructor
+  explicit TextValue(std::string name) : name(name) {}
+};
+
+// operator overloading
+inline std::ostream& operator<<(std::ostream& os, const TextValue& val) {  // NOLINT(*)
+  return os << val.name;
+}
+
+/*!
+ * \brief Meta data context for TextPrinter.
+ *
+ * This is an important part to enable bi-directional serializability.
+ * We use tvm's Node system to build the current IR.
+ * It can hard to design a text format for all the possible nodes
+ * as the set of nodes can grow when we do more extensions.
+ *
+ * Instead of trying to design readable text format for every nodes,
+ * we support a meta-data section in the text format.
+ * We allow the text format to refer to a node in the meta-data section.
+ *
+ * The meta-data section is a json serialized string of an Array<NodeRef>.
+ * Each element in the meta-data section can be referenced by the text format.
+ * Each meta data node is printed in the following format.
+ *
+ * meta.<type-key-of-node>(<index-in-meta-section>)
+ *
+ * Specifically, consider the following IR(constructed by python).
+ *
+ * \code
+ *
+ * n = tvm.var("n")
+ * x = tvm.relay.var("x", shape=(n, 1))
+ * f = tvm.relay.Function([x], x)
+ * print(f.astext())
+ *
+ * \endcode
+ *
+ * The corresponding text format is shown in the following code block.
+ *
+ * \code
+ *
+ * function(%x: Tensor[(meta.Variable(id=0),), float32]) {
```

Comment:
May I confirm, this line means `%x` is a float32 Tensor, whose shape is given by the first variable in the meta data?